assets/helpers/paymentIntegrations/paymentApi.js

```diff
@@ -0,0 +1,240 @@
+import { type Participations } from 'helpers/abTests/abtest';
+import { type UsState, type CaState, type IsoCountry } from 'helpers/internationalisation/country';
+import { pollUntilPromise, logPromise } from 'helpers/promise';
+import trackConversion from 'helpers/tracking/conversions';
+
+import * as cookie from 'helpers/cookie';
+
+// ----- Types ----- //
+
+type OneOffFields = {|
+  paymentData: {
+    currency: IsoCurrency,
+    amount: number,
+    token: string,
+    email: string
+  },
+  acquisitionData: PaymentAPIAcquisitionData,
+|};
+
+type RegularContribution = {|
+  amount: number,
+  currency: string,
+  billingPeriod: BillingPeriod,
+|};
+
+type PayPalDetails = {| baid: string |};
+
+type StripeDetails = {| stripeToken: string |};
+
+type PaymentDetails = PayPalDetails | StripeDetails;
+
+type RegularFields = {|
+  firstName: string,
+  lastName: string,
+  country: IsoCountry,
+  state: UsState | CaState | null,
+  email: string,
+  contribution: RegularContribution,
+  paymentFields: PaymentDetails,
+  ophanIds: OphanIds,
+  referrerAcquisitionData: ReferrerAcquisitionData,
+  supportAbTests: AcquisitionABTest[],
+|};
+
+export type PaymentFields
+  = {| contributionType: 'oneoff', fields: OneOffFields |}
+  | {| contributionType: 'regular', fields: RegularFields |};
+
+type Credentials = 'omit' | 'same-origin' | 'include';
+
+export type Token
+  = {| paymentMethod: 'Stripe', token: string |}
+  | {| paymentMethod: 'PayPal', token: string |}
+  | {| paymentMethod: 'DirectDebit', accountHolderName: string, sortCode: string, accountNumber: string |};
+
+export type PaymentResult
+  = {| paymentStatus: 'success' |}
+  | {| paymentStatus: 'failure', error: string |};
+
+export type PaymentCallback = Token => Promise<PaymentResult>;
+
+// ----- Setup ----- //
+
+const PaymentSuccess: PaymentResult = { paymentStatus: 'success' };
+const POLLING_INTERVAL = 3000;
+const MAX_POLLS = 10;
+const ONEOFF_CONTRIB_ENDPOINT = window.guardian.paymentApiStripeEndpoint;
+
+// ----- Functions ----- //
+
+/** Builds a `RequestInit` object for use with GET requests using the Fetch API */
+function getRequestOptions(
+  credentials: Credentials,
+  csrf: CsrfState | null,
+): Object {
+  const headers = csrf !== null
+    ? { 'Content-Type': 'application/json', 'Csrf-Token': csrf.token || '' }
+    : { 'Content-Type': 'application/json' };
+
+  return {
+    method: 'GET',
+    headers,
+    credentials,
+  };
+}
+
+/** Builds a `RequestInit` object for use with POST requests using the Fetch API */
+function postRequestOptions(
+  data: PaymentFields,
+  credentials: Credentials,
+  csrf: CsrfState | null,
+): Object {
+  return { ...getRequestOptions(credentials, csrf), method: 'POST', body: JSON.stringify(data.fields) };
+}
+
+/** Sends a request to the payment API (or support workers, yolo) and converts
+ *  the response into a JSON object */
+function requestPaymentApi(endpoint: string, settings: Object) {
```

Comment:
This function is actually very general so I think a more general name would be more appropriate, e.g. `fetchJsonWithErrorLogging`, and it could also be moved under `assets/helpers`. That would avoid using `PaymentApi` in the name which is misleading